File: patterns/01-design-patterns/02-structural/01-adapter/js/adapter.mjs
export class ProviderError extends Error {
  constructor(code, message) {
    super(message);
    this.code = code;
  }
}

export class NormalizedError extends Error {
  constructor(code, retryable, message) {
    super(message);
    this.code = code;
    this.retryable = retryable;
  }
}

export class LegacyProviderAdapter {
  constructor(client) {
    this.client = client;
  }

  async complete(request) {
    if (!request || !Array.isArray(request.messages)) {
      throw new NormalizedError(
        "invalid_request",
        false,
        "request.messages must be an array",
      );
    }
    if (request.requiresTools === true) {
      throw new NormalizedError(
        "unsupported_capability",
        false,
        "legacy provider does not support tools",
      );
    }

    const deployment = { "chat-pro": "legacy-chat-pro" }[request.model];
    if (deployment === undefined) {
      throw new NormalizedError(
        "invalid_request",
        false,
        "unknown model: " + request.model,
      );
    }

    const legacyRequest = {
      deployment,
      prompt: request.messages
        .map((message) => message.role + ": " + message.content)
        .join("\n"),
    };

    let legacyResponse;
    try {
      legacyResponse = await this.client.generate(legacyRequest);
    } catch (error) {
      throw normalizeError(error);
    }

    return {
      content: legacyResponse.output,
      finishReason:
        { stop: "stop", length: "length" }[legacyResponse.stopCode] ?? "unknown",
    };
  }
}

function normalizeError(error) {
  if (!(error instanceof ProviderError)) {
    return new NormalizedError("upstream_failure", true, error.message);
  }
  if (error.code === "OVER_QUOTA") {
    return new NormalizedError("rate_limited", true, error.message);
  }
  if (error.code === "BAD_REQUEST") {
    return new NormalizedError("invalid_request", false, error.message);
  }
  return new NormalizedError("upstream_failure", true, error.message);
}

File: patterns/01-design-patterns/02-structural/01-adapter/python/adapter.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol, Sequence


@dataclass(frozen=True)
class Message:
    role: str
    content: str


@dataclass(frozen=True)
class ChatRequest:
    model: str
    messages: Sequence[Message]
    requires_tools: bool = False


@dataclass(frozen=True)
class ChatResponse:
    content: str
    finish_reason: str


@dataclass(frozen=True)
class LegacyRequest:
    deployment: str
    prompt: str


@dataclass(frozen=True)
class LegacyResponse:
    output: str
    stop_code: str


class ProviderError(Exception):
    def __init__(self, code: str, message: str) -> None:
        super().__init__(message)
        self.code = code


class NormalizedError(Exception):
    def __init__(self, code: str, retryable: bool, message: str) -> None:
        super().__init__(message)
        self.code = code
        self.retryable = retryable


class LegacyClient(Protocol):
    def generate(self, request: LegacyRequest) -> LegacyResponse: ...


class LegacyProviderAdapter:
    _MODEL_MAP = {"chat-pro": "legacy-chat-pro"}
    _STOP_MAP = {"stop": "stop", "length": "length"}

    def __init__(self, client: LegacyClient) -> None:
        self._client = client

    def complete(self, request: ChatRequest) -> ChatResponse:
        if request.requires_tools:
            raise NormalizedError(
                "unsupported_capability",
                False,
                "legacy provider does not support tools",
            )

        deployment = self._MODEL_MAP.get(request.model)
        if deployment is None:
            raise NormalizedError(
                "invalid_request",
                False,
                f"unknown model: {request.model}",
            )

        legacy_request = LegacyRequest(
            deployment=deployment,
            prompt="\n".join(
                f"{message.role}: {message.content}" for message in request.messages
            ),
        )

        try:
            legacy_response = self._client.generate(legacy_request)
        except ProviderError as error:
            raise self._normalize_error(error) from error

        return ChatResponse(
            content=legacy_response.output,
            finish_reason=self._STOP_MAP.get(legacy_response.stop_code, "unknown"),
        )

    @staticmethod
    def _normalize_error(error: ProviderError) -> NormalizedError:
        if error.code == "OVER_QUOTA":
            return NormalizedError("rate_limited", True, str(error))
        if error.code == "BAD_REQUEST":
            return NormalizedError("invalid_request", False, str(error))
        return NormalizedError("upstream_failure", True, str(error))
